File: app/src/main/java/com/falcon/restaurants/screens/restaurant/RestaurantsActivity.java
package com.falcon.restaurants.screens.restaurant;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ImageView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.falcon.restaurants.R;
import com.falcon.restaurants.base.LocalListeners;
import com.falcon.restaurants.screens.common.BaseActivity;
import com.falcon.restaurants.screens.common.ImageLoader;
import com.falcon.restaurants.screens.common.ScreensNavigator;
import com.falcon.restaurants.screens.common.dialogs.DialogsNavigator;
import com.falcon.restaurants.utils.Logger;

import javax.inject.Inject;

public class RestaurantsActivity extends BaseActivity {
    String TAG = "RestaurantsActivity" ;

    private String parentId = "0";
    public static String PARENT_ID = "PARENT_ID" ;

    @Inject public RestaurantViewModel restaurantViewModel;
    @Inject public ScreensNavigator screensNavigator ;
    @Inject public DialogsNavigator dialogsNavigator;
    @Inject public LayoutInflater layoutInflater;
    @Inject public ImageLoader imageLoader;
    @Inject public FilterRestaurantsUseCase filterRestaurantsUseCase;

    RestaurantsListAdapter adapter ;
    SearchView searchView ;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        getPresentationComponent().inject(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_restaurants);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        RecyclerView recyclerView = findViewById(R.id.recyclerview);

        adapter = new RestaurantsListAdapter(
                this,
                layoutInflater,
                filterRestaurantsUseCase,
                imageLoader);

        adapter.setOnRestaurantClickListener(new RestaurantsListAdapter.OnRestaurantClickListener() {
            @Override
            public void onRestaurantClicked(String restaurantId) {
                launchProperActivity(restaurantId);
            }
        });

        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        Intent intent = getIntent();
        if(intent.hasExtra(RestaurantsActivity.PARENT_ID)){
            parentId = intent.getStringExtra(RestaurantsActivity.PARENT_ID);
        }
        Logger.log( TAG,"onCreate: parentId: " + parentId);

        setAdapterList();
    }

    public void setAdapterList(){
        restaurantViewModel.getByParentId(parentId).observe(this, categories -> {
            Logger.log( TAG,"setAdapterList: called");
            adapter.setMainList(categories);
        });
    }

    private void launchProperActivity(String restaurantId) {
        restaurantViewModel.hasChildren(restaurantId, new LocalListeners.OnSuccessListener() {
            @Override
            public void onSuccess(Boolean hasChildren) {
                Logger.log( TAG, "onSuccess: hasChildren: " + hasChildren);
                if(hasChildren){
                    screensNavigator.toCategoriesActivity(restaurantId);
                }else {
                    screensNavigator.toMealsActivity(restaurantId);
                }
            }
            @Override
            public void onFailed() {
                Logger.log( TAG, "onFailed: ");
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        final MenuItem myActionMenuItem = menu.findItem( R.id.mnuSearch);
        searchView = (SearchView) myActionMenuItem.getActionView();
        ImageView searchIcon = searchView.findViewById(R.id.search_button);
        searchIcon.setImageDrawable(ContextCompat.getDrawable(this,R.drawable.baseline_search_white_24));
        searchView.setMaxWidth(Integer.MAX_VALUE);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                 //Logger.log( "onQueryTextSubmit: ");
                if( ! searchView.isIconified()) {
                    searchView.setIconified(true);
                }
                myActionMenuItem.collapseActionView();
                return false;
            }
            @Override
            public boolean onQueryTextChange(String s) {
                //Logger.log( "onQueryTextChange: ");
                adapter.filter(s);
                return false;
            }
        });

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        int id = item.getItemId();
        return super.onOptionsItemSelected(item);
    }

    public static void start(AppCompatActivity fromActivity, String restaurantId){
        Intent intent = new Intent(fromActivity, RestaurantsActivity.class);
        intent.putExtra(RestaurantsActivity.PARENT_ID, restaurantId);
        fromActivity.startActivity(intent);
        if(restaurantId.equals("0")) {
            fromActivity.finish();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        return true ;
    }
}

File: app/src/main/java/com/falcon/restaurants/screens/restaurant/RestaurantsActivity.kt
package com.falcon.restaurants.screens.restaurant

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.falcon.restaurants.R
import com.falcon.restaurants.base.LocalListeners
import com.falcon.restaurants.room.restaurant.Restaurant
import com.falcon.restaurants.screens.common.BaseActivity
import com.falcon.restaurants.screens.common.ImageLoader
import com.falcon.restaurants.screens.common.ScreensNavigator
import com.falcon.restaurants.screens.common.dialogs.DialogsNavigator
import com.falcon.restaurants.utils.Logger

import javax.inject.Inject

class RestaurantsActivity : BaseActivity() {

    val TAG: String = "RestaurantsActivity" 

    var parentId: String = "0"

    @Inject lateinit var restaurantViewModel: RestaurantViewModel
    @Inject lateinit var screensNavigator: ScreensNavigator 
    //@Inject lateinit var dialogsNavigator: DialogsNavigator
    @Inject lateinit var layoutInflator: LayoutInflater
    @Inject lateinit var imageLoader: ImageLoader
    @Inject lateinit var filterRestaurantsUseCase: FilterRestaurantsUseCase

    lateinit var adapter: RestaurantsListAdapter 
    lateinit var searchView: SearchView 

    override fun onCreate(savedInstanceState: Bundle?) {
        presentationComponent.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_restaurants)
        val toolbar: Toolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)
        val recyclerView: RecyclerView = findViewById(R.id.recyclerview)

        adapter = RestaurantsListAdapter(
                this,
                layoutInflator,
                filterRestaurantsUseCase,
                imageLoader)

        adapter.onRestaurantClickListener = object: RestaurantsListAdapter.OnRestaurantClickListener{
            override fun onRestaurantClicked(restaurantId: String) {
                launchProperActivity(restaurantId)
            }
        }

        recyclerView.setAdapter(adapter)
        recyclerView.setLayoutManager(LinearLayoutManager(this))

        if(intent.hasExtra(PARENT_ID)){
            parentId = intent.getStringExtra(PARENT_ID)!!
        }
        Logger.log( TAG,"onCreate: parentId: " + parentId)

        setAdapterList()
    }

    fun setAdapterList(){
        restaurantViewModel.getByParentId(parentId).observe(this, {categories ->
                Logger.log(TAG, "setAdapterList: called")
                adapter.main_arrlst = categories as MutableList<Restaurant>
            })
    }

    fun launchProperActivity(restaurantId: String) {
        restaurantViewModel.hasChildren(restaurantId, object : LocalListeners.OnSuccessListener{
            override fun onSuccess(hasChildren: Boolean) {
                Logger.log( TAG, "onSuccess: hasChildren: " + hasChildren)
                if(hasChildren){
                    screensNavigator.toCategoriesActivity(restaurantId)
                }else {
                    screensNavigator.toMealsActivity(restaurantId)
                }
            }
            override fun onFailed() {
                Logger.log( TAG, "onFailed: ")
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        getMenuInflater().inflate(R.menu.menu_main, menu)
        val myActionMenuItem: MenuItem = menu.findItem( R.id.mnuSearch)
        searchView = myActionMenuItem.getActionView() as SearchView
        val searchIcon: ImageView = searchView.findViewById(R.id.search_button)
        searchIcon.setImageDrawable(ContextCompat.getDrawable(this,R.drawable.baseline_search_white_24))
        searchView.setMaxWidth(Integer.MAX_VALUE)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener{

            override fun onQueryTextSubmit(query: String): Boolean {
                 //Logger.log( "onQueryTextSubmit: ")
                if( ! searchView.isIconified()) {
                    searchView.setIconified(true)
                }
                myActionMenuItem.collapseActionView()
                return false
            }
            override fun onQueryTextChange(s: String): Boolean {
                //Logger.log( "onQueryTextChange: ")
                adapter.filter(s)
                return false
            }
        })

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id: Int = item.getItemId()
        return super.onOptionsItemSelected(item)
    }

    companion object {
        const val PARENT_ID: String = "PARENT_ID" 
        fun start(fromActivity: AppCompatActivity, restaurantId: String)
        {
            val intent: Intent = Intent(fromActivity, RestaurantsActivity::class.java)
            intent.putExtra(PARENT_ID, restaurantId)
            fromActivity.startActivity(intent)
            if (restaurantId.equals("0")) {
                fromActivity.finish()
            }
        }
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        return true 
    }
}